### PR TITLE
[Core] Add a `store_names` property to the bot, thats checked before storing username and nicknames as well as a command to delete currently stored names

### PR DESCRIPTION
**Pull request by Drapersniper**
Stop storing names by default and add toggle to store names and command to delete stored names

### Type

- [ ] Bugfix
- [x] Enhancement
- [x] New feature

### CODE REVIEW

**Comment by Dav-Git** [comment]
Why though?

**Comment by Drapersniper** [comment]
1 - Names and User names are User data, so to comply with discord you need to take some extra steps in handling these
2 - Some users may not wish to store names and usernames specially larger bots
3 - Some bot owners may not even be away that these are stored.

**Comment by SharkyTheKing** [comment]
I would love to see this, not everyone wants to store nickname/username information on the bot's members. For large bots/servers storing names adds extra load for the person/VPS hosting the bot, being able to toggle it off/on as a core functionality would be beneficial. Allowing the bot owner to decide if they're fine with storing names or not is great, especially since they'll be able to judge if they want that type of functionality or if they can handle that functionality alongside whatever else they're using.

**Comment by Drapersniper** [comment]
Okay ... so ... upon further inspection i hit some issues when doing the follow up PR so will be changing the implementation slighly on this one

**Comment by Drapersniper** [comment]
Okay because life is hard.....

I've added `Red.store_names` attribute to the bot,
This attribute is now checked before storing names in Bank, Modlog and Mod.

I've added the following commands:
`[p]bankset deletenames` to delete the usernames stored in Bank config
`[p]modset deletenames` to delete the usernames stored in Mod config
`[p]modlogset deletenames` to delete the usernames stored in ModLog config

`[p]set storenames` to toggle the state of this config that by default is off
`[p]set deletenames` to delete which delete the stored names and nicknames for all the 3 cogs we currently store names for.


since `bot.store_names` is now a property in the bot, CC can also use it going forward to know whether the bot owner wishes to store username/nicknames.

**Comment by Bakersbakebread** [comment]
This PR leaves much to be desired, but is definitely a step in the right direction. 

A user ID is not considered End User Data but a Name is, therefore I would like to see an implementation with use of this PR from bot devs/CC's where there is express consent given from the End User as per TOS. 

Having this opt-in is a **must** as the accountability falls on the implementer. We should not be storing EUD beyond the scope of the application, again per TOS. Names seems redundant in my opinion as this information is available by the API - any information returned by the API is also the sole responsibility of Discord and not bot devs/CC's respectively.

File: redbot/cogs/mod/settings.py [hunk 1]
```diff
@@ -221,3 +222,21 @@ async def defaultdays(self, ctx: commands.Context, days: int = 0):
                 days=days
             )
         )
+
+    @modset.command()
+    @commands.is_owner()
+    async def deletenames(self, ctx: commands.Context) -> None:
+        """Delete all stored names and usernames"""
+
+        async with self.settings._get_base_group(self.settings.MEMBER).all() as mod_member_data:
```
Comment (Bakersbakebread) [inline redbot/cogs/mod/settings.py]:
```
Traceback (most recent call last):
  File "a:\projects\venvs\testvenv\lib\site-packages\discord\ext\commands\core.py", line 83, in wrapped
    ret = await coro(*args, **kwargs)
  File "c:\users\seant\desktop\ruinedred\redbot\cogs\mod\settings.py", line 231, in deletenames
    async with self.settings._get_base_group(self.settings.MEMBER).all() as mod_member_data:
AttributeError: 'Mod' object has no attribute 'settings'
```

File: redbot/core/core_commands.py [hunk 2]
```diff
@@ -976,6 +979,70 @@ async def setdescription(self, ctx: commands.Context, *, description: str = ""):
...
+    @checks.is_owner()
+    @_set.command(name="storenames")
+    async def storenames(self, ctx: commands.Context):
+        """
+        Toggle whether or not to store user and nick names.
+
+        Default is Off
+        """
+        ctx.bot._store_names = not ctx.bot._store_names
+        await ctx.bot._config.store_names.set(ctx.bot.store_names)
```
Comment (Bakersbakebread) [inline redbot/core/core_commands.py:991]: A DateTime object is being passed here - can't trace where from? I had zero config setup

File: redbot/core/core_commands.py [hunk 3]
```diff
@@ -976,6 +979,70 @@ async def setdescription(self, ctx: commands.Context, *, description: str = ""):
...
+        if ctx.bot.store_names:
+            await ctx.send(
+                _("The bot will now store usernames and nicknames in Bank, ModLog and Mod cogs.")
+            )
+        else:
+            await ctx.send(
+                _(
+                    "The bot will no longer store usernames and nicknames in Bank, "
+                    "ModLog and Mod cogs.\n"
+                    "To delete all stored names use `[p]set deletenames"
```
Comment (Bakersbakebread) [inline redbot/core/core_commands.py]:
```suggestion
                    "To delete all stored names use `[p]set deletenames`"
```

File: redbot/cogs/modlog/modlog.py [hunk 4]
```diff
@@ -33,6 +34,18 @@ async def reapply_audittype_migration(self, ctx: commands.Context):
...
+    @modlogset.command(name="deletenames")
+    async def modlogset_deletenames(self, ctx: commands.Context):
+        """Delete all stored usernames."""
+        async with modlog._config.custom("CASES").all() as modlog_data:
+            for guild_id, guild_data in modlog_data.items():
+                for case_number, case_data in guild_data.items():
+                    if "last_known_username" in modlog_data[guild_id][case_number]:
+                        del modlog_data[guild_id][case_number]["last_known_username"]
+                    await asyncio.sleep(0)
+        await ctx.tick()
```
Comment (Bakersbakebread) [inline redbot/cogs/modlog/modlog.py]:
```suggestion
        await ctx.send(_("Usernames have been deleted from the ModLog config."))
```
This `ctx.tick()` is inconsistent with:
- `bankset deletenames`
- `modset deletenames`

**Review by Bakersbakebread — APPROVED** [review]
I'm not sure on other parts of the codebase, but I thought of maybe splitting out "last_known_username" into a constant somehere  ¯\\_(ツ)_/¯

**Comment by Drapersniper** [comment]
Closing in favour of the data api PR that has been merged

**Pull request closed without merging**